pkg/controller/nstemplateset/cluster_resources.go

```diff
@@ -90,73 +168,138 @@ func (r *clusterResourcesManager) ensure(logger logr.Logger, nsTmplSet *toolchai
 	// As a consequence, when the NSTemplateSet is deleted, we explicitly delete the associated cluster-wide resources that belong to the same user.
 	// see https://issues.redhat.com/browse/CRT-429
 
-	for _, obj := range newObjs {
-		logger.Info("applying cluster resources object", "obj", obj.Object.GetObjectKind(), "username", username)
-		if createdOrUpdated, err := applycl.NewApplyClient(r.client, r.scheme).Apply([]runtime.RawExtension{obj}, labels); err != nil {
-			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusClusterResourcesProvisionFailed, err, "failed to create cluster resources")
-		} else if createdOrUpdated {
-			namespaces, err := fetchNamespaces(r.client, username)
-			if err != nil {
-				return true, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err, "failed to list namespace with label owner '%s'", username)
+	logger.Info("applying cluster resource", "gvk", toApply.GetGvk())
+	if _, err := applycl.NewApplyClient(r.client, r.scheme).Apply([]applycl.ToolchainObject{toApply}, labels); err != nil {
+		return false, fmt.Errorf("failed to apply cluster resource of type '%v'", toApply.GetGvk())
+	}
+	return true, nil
+}
+
+// updateOrDeleteRedundant takes the given currentObjs and newObjs and compares them.
+//
+// If there is any existing redundant resource (exist in the currentObjs, but not in the newObjs), then it deletes the resource and returns 'true, nil'.
+//
+// If there is any resource that is outdated (exists in both currentObjs and newObjs but its templateref is not matching),
+// then it updates the resource and returns 'true, nil'
+//
+// If no resource to be updated or deleted was found then it returns 'false, nil'. In case of any errors 'false, error'
+func (r *clusterResourcesManager) updateOrDeleteRedundant(logger logr.Logger, currentObjs []applycl.ComparableToolchainObject, newObjs []applycl.ToolchainObject, tierTemplate *tierTemplate, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (bool, error) {
+	// go though all current objects so we can compare then with the set of the requested and thus update the obsolete ones or delete redundant ones
+	for _, currentObject := range currentObjs {
+
+		// if the template is not specified, then delete all cluster resources one by one
+		if nsTmplSet.Spec.ClusterResources == nil || tierTemplate == nil {
+			return r.deleteClusterResource(nsTmplSet, currentObject)
+		}
+
+		// if the existing object is not up-to-date, then check if it should be only updated or completely removed (in case it's missing the in set of the requested objects)
+		if !isUpToDateAndProvisioned(currentObject, tierTemplate) {
+			for _, newObject := range newObjs {
+				if newObject.HasSameName(currentObject) {
+					// is found then let's update it
+					if err := r.setStatusUpdatingIfNotProvisioning(nsTmplSet); err != nil {
+						return false, err
+					}
+					return r.apply(logger, nsTmplSet, tierTemplate, newObject)
+				}
 			}
```

Comment:
WDYT about something like this:
```suggestion
		for _, newObject := range newObjs {
			if newObject.HasSameName(currentObject) {
				// is found then let's check if we need to update it
				if !isUpToDate(currentObject, newObject, tierTemplate) {
					// let's update it
					if err := r.setStatusUpdatingIfNotProvisioning(nsTmplSet); err != nil {
						return false, err
					}
					return r.apply(logger, nsTmplSet, tierTemplate, newObject)
				}
			}
		}

...
// isUpToDate returns true if the currentObject uses the corresponding templateRef and equals to the new object
func isUpToDate(currentObject, newObject applycl.ToolchainObject, tierTemplate *tierTemplate) bool {
	return currentObject.GetLabels()[toolchainv1alpha1.TemplateRefLabelKey] != "" &&
		currentObject.GetLabels()[toolchainv1alpha1.TemplateRefLabelKey] == tierTemplate.templateRef
		// && currentObject.IsSame(newObject)  <-- TODO Uncomment wheh IsSame is implemented for all ToolchainObjects!
}

```

Reply:
changed in ef2a0df